utilities/Microsoft.Quantum.Katas/KataMagic.cs

```diff
@@ -228,5 +233,138 @@ public virtual OperationInfo FindSkeletonAnswer(OperationInfo test, OperationInf
+                        // https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.gettype?view=net-5.0#System_Reflection_Assembly_GetType_System_String_
+                        Type? simType = asm.GetType(simTypeName);
+
+                        // Using binding flags to invoke parameterised constructor with default values
+                        // For more details, please refer https://stackoverflow.com/questions/2501143/activator-createinstancetype-for-a-type-without-parameterless-constructor
+                        simulator = (simType != null)
+                            ?   Activator.CreateInstance
+                                (
+                                    simType,
+                                    BindingFlags.CreateInstance |
+                                    BindingFlags.Public |
+                                    BindingFlags.Instance |
+                                    BindingFlags.OptionalParamBinding,
+                                    null,
+                                    null,
+                                    null
+                                )
+                            :   null;
+
+                        if(simulator != null)
+                            break;
+                    }
+                    string errorMessage = $"Error while creating an instance of {simName}. " +
+                        $"{simName} not a valid execution target. " +
+                        "Either consider using a fully qualified name for the simulator. " +
+                        "or see that you aren't using Custom simulators as part of Q# projects." ;
+                    if(simulator == null)
+                        throw new Exception(errorMessage);
+                }
+                catch (Exception ex)
+                {
+                    channel.Stderr(ex.Message);
+                    throw new Exception("Error while creating desired simulator(s) for test.");
+                }
+
+                if(simulator is QuantumSimulator qsim)
+                {
+                    // WithJupyterDisplay() adds functionality to a given
+                    // quantum simulator to display diagnostic output
+                    // with rich Jupyter formatting
+                    testSimulators.Add(qsim.WithJupyterDisplay(channel, ConfigurationSource));
+                }
+                else if(simulator is SimulatorBase sim)
+                {
+                    testSimulators.Add( sim.WithStackTraceDisplay(channel) );
+                    // Add logging ability for simulators other than quantum simulator
+                    sim.OnLog += channel.Stdout;
+                }
+
+            }
+            return testSimulators;
+        }
+
+        /// <summary>
+        /// Returns the list of simulator names specified via @Test attribute
+        /// </summary>
+        public List<string> GetSimNamesFromTestAttribute(OperationInfo test)
```

Comment:
It would be good to avoid the duplicated code between KataMagic and CheckKataMagic where possible. Creating a shared superclass would be a great way to reduce the duplication.

Reply:
Great idea. I was thinking of `UtilityMagic` as a name for the superclass. Does that work or do you have another name in mind?

Reply:
`CheckKataMagic` and `KataMagic` are related but there isn't a clear hierarchical relationship, so best to give them a common parent. Let's follow the same pattern used in iqsharp and name it `AbstractKataMagic` (see `AbstractMagic` from iqsharp [here](https://github.com/microsoft/iqsharp/blob/main/src/Jupyter/Magic/AbstractMagic.cs)).

Reply:
Sounds great

Reply:
Done the refractoring